python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -70,6 +72,95 @@ def cuda_array_interface_wrapper(
     )
 
 
+class BufferWeakref(object):
```

Comment:
Using `CachedinstanceMeta` or something very similar to it, we should _enforce_ that no two instances of `BufferWeakref` can exist with the same `size` and `ptr` attributes.